### ISSUE TITLE
The clustering algorithm seem to be less useful

### ISSUE THREAD

**Issue by songyukino1** [theRenard/cocos-creator-tiled-collider-creator#1]
Thank you very much for sharing, which inspired me. However, in practice, I found that this clustering algorithm seems to be less useful on my map. I tried adjusting the parameters, but it still brings together some points that should not be grouped together.

**Comment by theRenard** [comment]
Hi @songyukino1 it could happen when you have 1 unit between two platforms. ie: 

will not work
```
____ ____
```
will work
```
____  ____
```
This is true for x and y direction so for this to work the compromise is to put some extra space between your points when needed.

I was working on another project create a plugin for DLtk but never found the time to finish. In that case the algorithm is very good. I could try to create a v2 for this with the new algorithm that could work

**Comment by songyukino1** [comment]
I found that there is support for composite collision in TiledMap in Unity. How do you think it is implemented?

**Comment by songyukino1** [comment]
I just want to avoid getting stuck in the gaps between tiles while the character is moving. Do you think there is a simpler way?

**Comment by PatrickKunz** [comment]
This algorithm also didn't work for me. IMHO this is the wrong approach and overcomplicated. Too much can go wrong. 

You can use the polygon-clipping npm package and union all your tile polygons in one step.

**Comment by theRenard** [comment]
Yeah, sure it is overcomplicated! Not sure if I'm gonna develop for Cocos in the near future but I was exploring LDtk editor and I wrote a better algorithm based on polybool, you can find it [here](https://github.com/theRenard/Test_DLtk/tree/master), not sure if that can help because I've never found the time to adapt it to Tiled, you might have better luck 😓

![image](https://github.com/theRenard/cocos-creator-tiled-collider-creator/assets/61255863/2178a5bb-a57a-43ea-a986-9f666e624e96)
![image](https://github.com/theRenard/cocos-creator-tiled-collider-creator/assets/61255863/e46a8872-a134-469f-94af-e08caafe0c5c)
![image](https://github.com/theRenard/cocos-creator-tiled-collider-creator/assets/61255863/9f5300ca-c500-494d-a717-fb38aab211ac)

**Comment by PatrickKunz** [comment]
> not sure if that can help

Thanks for getting back and thanks for sharing your solution. It pointed me in the right direction. I found something that works. 

We have tile polygons that touch each other. Because of this, we can use a normal union algorithm, like vector drawing programs have, to merge the tiles. The npm package polygon-clipping contains such an algorithm and it is fast. 

No special magic numbers or tuning is required. It just works. 

The code looks like this:

```
  import polygonClipping, { Polygon } from "polygon-clipping";
....

   start() {
      const tiledLayer = this.node.getComponent(TiledLayer);
      const { width, height } = tiledLayer.getMapTileSize();
      const { leftDownToCenterX, leftDownToCenterY } = tiledLayer;

      console.log("Tile Size: " + width + " " + height);

      const vertices = tiledLayer.vertices;
      const polygons: polygonClipping.MultiPolygon = [];

      for (const vertex of vertices) 
      {
        if (vertex === undefined) 
        {
          continue;
        }

        const entries = Object.entries(vertex);
        for (const [key, value] of entries) 
        {
          if (typeof value === "object")
          {
            const { left, bottom } = value;

            let x = left - leftDownToCenterX;
            let y = bottom - leftDownToCenterY;

            let tileVertices : polygonClipping.Polygon = 
              [[[x, y],
              [x + width, y],
              [x + width, y + height],
              [x, y + height],
              [x, y]]]; // close path

            polygons.push(tileVertices);
          }
        }
      }

      const polygonResult : polygonClipping.MultiPolygon = polygonClipping.union(polygons);

      for (let i = 0; i < polygonResult.length; i++)
      {
        const collider = this.node.addComponent(PolygonCollider2D);
        collider.points = polygonResult[i][0].map(([x, y]) => new Vec2(x, y));
        collider.apply();
      }
    }

```
